[TRACEPOINT] Add sync object and skip non-LIVE modules

Adam-pi3 opened the pull request:
TRACEPOINT support has assumption that function
'p_arch_static_call_transform_entry' can't run concurrently.
However, mkosi hit the problem that is not always the case.
This commit should address described issue tracked by #87

### Description
printk() at line 60 is for debugging. If we see this message in mkosi logs, it means we hit the problem

### How Has This Been Tested?
<!--- Please describe how you tested your changes. -->

solardiz commented:
Looks like we happened to hit a variation of #87 on "mkosi boot (mainline) / mkosi-mainline (pull_request) Failing after 8m — mkosi-mainline" this time. I guess this is unrelated to this PR's changes.

```
[  OK  ] Started Journal Service.
         Starting Create Volatile Files and Directories...
[  OK  ] Finished Create Volatile Files and Directories.
[  OK  ] Finished dracut cmdline hook.
         Starting dracut pre-udev hook...
[    8.543682] p_lkrg: loading out-of-tree module taints kernel.
[    8.544701] p_lkrg: module verification failed: signature and/or required key missing - tainting kernel
[    8.548503] [p_lkrg] Loading LKRG...
[    8.548975] [p_lkrg] System does NOT support SMEP. LKRG can't enforce SMEP validation :(
[    8.549386] [p_lkrg] System does NOT support SMAP. LKRG can't enforce SMAP validation :(
[    8.621421] Freezing user space processes ... (elapsed 0.001 seconds) done.
[    8.625156] OOM killer disabled.
[    9.768551] [p_lkrg] [kretprobe] register_kretprobe() for <ovl_create_or_link> failed! [err=-2]
[    9.769006] [p_lkrg] Can't hook 'ovl_create_or_link' function. This is expected if you are not using OverlayFS.
[    9.896299] BUG: kernel NULL pointer dereference, address: 0000000000000000
[    9.899171] #PF: supervisor instruction fetch in kernel mode
[    9.899171] #PF: error_code(0x0010) - not-present page
[    9.899171] PGD 0 P4D 0 
[    9.899171] Oops: 0010 [#1] SMP NOPTI
[    9.899171] CPU: 0 PID: 12 Comm: rcu_sched Tainted: G           OE     5.13.0-051300rc1daily20210511-generic #202105102210
[    9.899171] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.13.0-1ubuntu1.1 04/01/2014
[    9.899171] RIP: 0010:0x0
[    9.899171] Code: Unable to access opcode bytes at RIP 0xffffffffffffffd6.
[    9.899171] RSP: 0018:ffffb8a6c006fe10 EFLAGS: 00000092
[    9.899171] RAX: 0000000000000000 RBX: 0000000000000000 RCX: 000000000e400000
[    9.899171] RDX: 000000000e400000 RSI: 0000000000000246 RDI: ffff90fdbec1cc00
[    9.899171] RBP: ffffb8a6c006fe68 R08: 00000000fffee437 R09: 00000000fffee437
[    9.899171] R10: 0000000000000000 R11: 0000000000000001 R12: 00000000fffee438
[    9.899171] R13: ffffb8a6c006fe20 R14: ffffffff93556f40 R15: ffff90fd8126ae00
[    9.899171] FS:  0000000000000000(0000) GS:ffff90fdbec00000(0000) knlGS:0000000000000000
[    9.899171] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    9.899171] CR2: ffffffffffffffd6 CR3: 0000000035490000 CR4: 00000000000006f0
[    9.899171] Call Trace:
[    9.899171]  elfcorehdr_read+0x40/0x40
[    9.899171]  ? __bpf_trace_tick_stop+0x10/0x10
[    9.899171]  rcu_gp_fqs_loop+0xe5/0x280
[    9.899171]  rcu_gp_kthread+0x85/0xf0
[    9.899171]  kthread+0x12f/0x150
[    9.899171]  ? rcu_gp_init+0x4d0/0x4d0
[    9.899171]  ? __kthread_bind_mask+0x70/0x70
[    9.899171]  ret_from_fork+0x22/0x30
[    9.899171] Modules linked in: p_lkrg(OE+) dm_mirror dm_region_hash dm_log virtio_rng autofs4
[    9.899171] CR2: 0000000000000000
ABORT
[    9.899171] ---[ end trace 7ff1958ee4a1aca5 ]---
[    9.899171] RIP: 0010:0x0
[    9.899171] Code: Unable to access opcode bytes at RIP 0xffffffffffffffd6.
[    9.899171] RSP: 0018:ffffb8a6c006fe10 EFLAGS: 00000092
[    9.899171] RAX: 0000000000000000 RBX: 0000000000000000 RCX: 000000000e400000
[    9.899171] RDX: 000000000e400000 RSI: 0000000000000246 RDI: ffff90fdbec1cc00
[    9.899171] RBP: ffffb8a6c006fe68 R08: 00000000fffee437 R09: 00000000fffee437
[    9.899171] R10: 0000000000000000 R11: 0000000000000001 R12: 00000000fffee438
[    9.899171] R13: ffffb8a6c006fe20 R14: ffffffff93556f40 R15: ffff90fd8126ae00
[    9.899171] FS:  0000000000000000(0000) GS:ffff90fdbec00000(0000) knlGS:0000000000000000
[    9.899171] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    9.899171] CR2: ffffffffffffffd6 CR3: 0000000035490000 CR4: 00000000000006f0
[    9.899171] Kernel panic - not syncing: Fatal exception
[    9.899171] Kernel Offset: 0x10600000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[    9.899171] ACPI MEMORY or I/O RESET_REG.
```

solardiz commented:
> This commit should address described issue tracked by #87

Adam, you must have meant #85. I could edit the GitHub comment, but the commit message is yours to fix - to do prior to possibly merging this.

solardiz commented on src/modules/database/TRACEPOINT/p_arch_static_call_transform/p_arch_static_call_transform.c:
```diff
@@ -40,15 +40,33 @@ static struct module *p_module2;
 static unsigned int p_module1_idx;
 static unsigned int p_module2_idx;
 
+p_lkrg_counter_lock p_static_call_spinlock;
```
`static`

solardiz commented on src/modules/database/TRACEPOINT/p_arch_static_call_transform/p_arch_static_call_transform.c:
```diff
@@ -40,15 +40,33 @@ static struct module *p_module2;
+         p_lkrg_counter_lock_unlock(&p_static_call_spinlock, &p_flags);
+         break;
+      }
+   } while(1);
```
I didn't review this for actually doing what we want here, but technically I'd simplify the code to:

```c
   do {
      p_lkrg_counter_lock_lock(&p_static_call_spinlock, &p_flags);
      if (p_lkrg_counter_lock_val_read(&p_static_call_spinlock)) {
         p_lkrg_counter_lock_unlock(&p_static_call_spinlock, &p_flags);
         cpu_relax();
         printk(KERN_CRIT "WAITING!\n");
         continue;
      }
   } while(0);
   p_lkrg_counter_lock_val_inc(&p_static_call_spinlock);
   p_lkrg_counter_lock_unlock(&p_static_call_spinlock, &p_flags);
```

(Avoid the need for the `else` clause.)

Also, if we're going to keep the `printk` in code to be merged into the main branch, we should probably put a more verbose message in there at least indicating that this message comes from LKRG.

solardiz reviewed (changes requested):
At least the commit message must be corrected not to refer to the wrong issue. I've also suggested code cleanups.

Adam-pi3 commented:
@solardiz  What about dropping this PR and i will push the changes directly to the main branch? The reason why is because when I changed the commit message (to fix #87 to #85) git is not happy. I changed that via `git commit --amend`.

solardiz commented:
>  What about dropping this PR and i will push the changes directly to the main branch?

I don't mind.

> The reason why is because when I changed the commit message (to fix #87 to #85) git is not happy. I changed that via `git commit --amend`.

That's puzzling. In what way is it not happy?

Adam-pi3 commented:
> That's puzzling. In what way is it not happy?

I get following problem with refs:

```
...
To https://github.com/Adam-pi3/lkrg
 ! [rejected]        main -> main (non-fast-forward)
error: failed to push some refs to 'https://github.com/Adam-pi3/lkrg'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

It's easier for me just to drop it than looking into it and fixing it :)

solardiz commented:
@Adam-pi3 Those messages merely mean that after having amended a previously pushed commit you (obviously) need to force-push, `git push -f`. This is how it works, no problem there.
